packages/devtools/src/schema/compiler.ts

```diff
@@ -0,0 +1,328 @@
+      }
+      object.properties[key] = value.schema
+      refs = [...refs, ...value.refs]
+    }
+
+    if (required.length !== 0) {
+      object.required = required
+    }
+
+    const schemaWithRefs = { schema: object, refs }
+    this.#refs[name] = schemaWithRefs
+
+    return schemaWithRefs
+  }
+
+  _compileList(
+    objectName: string,
+    fieldName: string,
+    definition: ListFieldDefinition
+  ): SchemaWithRefs {
+    const list: JSONSchema.Array = { type: 'array', maxItems: definition.maxLength }
+    if (definition.minLength != null) {
+      list.minItems = definition.minLength
+    }
+
+    let item: SchemaWithRefs | void
+    switch (definition.item.type) {
+      case 'enum':
+        item = this._compileEnum(objectName, fieldName, definition.item)
+        break
+      case 'object':
+        item = this._compileObjectReference(objectName, fieldName, definition.item)
+        break
+      case 'scalar':
+        item = this._compileScalar(definition.item)
+        break
+    }
+
+    if (item == null) {
+      throw new Error(`Could not compile item schema for list ${fieldName} of object ${objectName}`)
+    }
+    list.items = item.schema
+
+    return { schema: list, refs: item.refs }
+  }
+
+  _compileObjectReference(
+    objectName: string,
+    fieldName: string,
+    definition: ObjectReferenceFieldDefinition
+  ): SchemaWithRefs {
+    if (this.#src.models[definition.name] != null) {
+      throw new Error(
+        `Unsupported reference to model ${definition.name} in field ${fieldName} of object ${objectName}. References can only be made to embedded objects.`
+      )
+    }
+
+    const target = this.#src.objects[definition.name]
+    if (target == null) {
+      throw new Error(
+        `Missing object ${definition.name} referenced in field ${fieldName} of object ${objectName}`
+      )
+    }
+
+    // Ensure object is compiled and injected to definitions record
+    this._compileEmbedObject(definition.name, target)
+
+    return { schema: { $ref: `#/$defs/${definition.name}` }, refs: [definition.name] }
+  }
+
+  _compileEnum(
+    objectName: string,
+    fieldName: string,
+    definition: EnumFieldDefinition
+  ): SchemaWithRefs {
+    const values = this.#src.enums[definition.name]
+    if (values == null) {
+      throw new Error(
+        `Missing enum ${definition.name} referenced in field ${fieldName} of object ${objectName}`
+      )
+    }
+    if (this.#refs[definition.name] == null) {
+      this.#refs[definition.name] = {
+        schema: { type: 'string', title: definition.name, enum: values },
+        refs: [],
+      }
+    }
+    return { schema: { $ref: `#/$defs/${definition.name}` }, refs: [definition.name] }
+  }
+
+  _compileScalar(definition: ScalarFieldDefinition): SchemaWithRefs {
+    const title = definition.schema.title
+    // Scalars without title or that have properties changed from the defaults are injected directly as they are not reusable
```

Comment:
What does "properties changed from the defaults" mean in this case? Does this mean that there can be two different scalar values with the same title but different properties? Maybe it would be good to throw an error in such case?

Reply:
Yes exactly, but why throw an error? It's not a problem to use the same scalar with different validation rules, it's just that we can't reuse the JSON schema as if they were the same.